static/js/util/coupon.js

```diff
@@ -0,0 +1,85 @@
+// @flow
+import {
+  COUPON_CONTENT_TYPE_PROGRAM,
+  COUPON_CONTENT_TYPE_COURSE,
+  COUPON_CONTENT_TYPE_COURSERUN,
+  COUPON_AMOUNT_TYPE_PERCENT_DISCOUNT,
+  COUPON_AMOUNT_TYPE_FIXED_DISCOUNT,
+} from '../constants';
+import type {
+  Coupons,
+  Coupon,
+  CalculatedPrices,
+} from '../flow/couponTypes';
+import type {
+  CoursePrice,
+  CoursePrices,
+  Dashboard,
+} from '../flow/dashboardTypes';
+
+export const calculatePrices = (programs: Dashboard, prices: CoursePrices, coupons: Coupons): CalculatedPrices => {
+  let couponLookup: Map<number, Coupon> = new Map();
+  let priceLookup: Map<number, CoursePrice> = new Map();
+  for (const coupon of coupons) {
+    couponLookup.set(coupon.program_id, coupon);
+  }
+  for (const price of prices) {
+    priceLookup.set(price.program_id, price);
+  }
```

Comment:
Could also make use potentially of `new Map([ [id, obj], ... ])` as well - that might be more elegant than what I suggested.